### PR TITLE
ENG-14728: A new API to parameterize a SqlNode tree

### PR DESCRIPTION
**Pull request by Chaozz**
The change required for adding A new API to parameterize a SqlNode tree. keep updating.
 
- add setOperand() to SqlOrderBy
- make `SqlParserPos.startsBefore` public

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java [hunk 1]
```diff
@@ -38,10 +38,10 @@
     }
   };
 
-  public final SqlNode query;
-  public final SqlNodeList orderList;
-  public final SqlNode offset;
-  public final SqlNode fetch;
+  public SqlNode query;
```
Comment (kaikaizi) [inline core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: I prefer having getters/setters to achieve this. (If the outside caller does not need to assign a new value, you should not provide setter, as getter is sufficient for updating).

File: core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java [hunk 2]
```diff
@@ -68,6 +68,25 @@ public SqlOperator getOperator() {
...
+      orderList = (SqlNodeList) operand;
+      break;
+    case 2:
+      offset = operand;
+      break;
+    case 3:
+      fetch = operand;
+      break;
+    default:
+      throw new AssertionError(i);
```
Comment (kaikaizi) [inline core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: Detailed reason for failure would be helpful here.

Comment (Chaozz) [reply core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: I try to follow the pattern other `SqlCall` implementing `setOperand `. Refer to https://github.com/yzhang1991/calcite/blob/9b2249f969f096037861745d09d3e1064d3a30f4/core/src/main/java/org/apache/calcite/sql/SqlSelect.java#L96

File: core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java (hunk 1 above)
Comment (kaikaizi) [reply core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: My bad, saw the red `public final` as `private final`.

File: core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java [hunk 3]
```diff
@@ -68,6 +68,25 @@ public SqlOperator getOperator() {
     return ImmutableNullableList.of(query, orderList, offset, fetch);
   }
 
+  @Override public void setOperand(int i, SqlNode operand) {
```
Comment (yzhang1991) [inline core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: Please comment.

Comment (yzhang1991) [reply core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: From reading the code I think the `int i` should be an enum type.
The caller code is not readable passing a magic number `i`.

File: core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java (hunk 1 above)
Comment (yzhang1991) [reply core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: What is `fetch`?

File: core/src/main/java/org/apache/calcite/sql/parser/SqlParserPos.java [hunk 4]
```diff
@@ -276,7 +276,7 @@ public boolean overlaps(SqlParserPos pos) {
         || pos.startsBefore(this) && pos.endsAfter(this);
   }
 
-  private boolean startsBefore(SqlParserPos pos) {
+  public boolean startsBefore(SqlParserPos pos) {
```
Comment (yzhang1991) [inline core/src/main/java/org/apache/calcite/sql/parser/SqlParserPos.java]: Who uses this?

File: core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java (hunk 1 above)
Comment (Chaozz) [reply core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: > What is `fetch`?

`fetch` is for `limit x`

File: core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java (hunk 3 above)
Comment (Chaozz) [reply core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: I try to follow the same pattern `SqlSelect` implementing `setOperand `. Refer to https://github.com/yzhang1991/calcite/blob/9b2249f969f096037861745d09d3e1064d3a30f4/core/src/main/java/org/apache/calcite/sql/SqlSelect.java#L96
It is a method declared in the base class `sqlcall` but not implemented.

File: core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java (hunk 1 above)
Comment (yzhang1991) [reply core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: Add to a comment, thanks!

File: core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java (hunk 3 above)
Comment (yzhang1991) [reply core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: OK. Then just add a javadoc comment.

Comment (Chaozz) [reply core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: done

File: core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java (hunk 1 above)
Comment (Chaozz) [reply core/src/main/java/org/apache/calcite/sql/SqlOrderBy.java]: done

**Pull request merged**